### PR TITLE
Update Autorest C# - DigitalTwins

### PR DESCRIPTION
**Pull request by pakrym**
Contributes to https://github.com/Azure/azure-sdk-for-net/issues/13511

Major features

1. Collections are now always initialized and collection properties are readonly by default
2. Internal deserialization ctors are removed for input-only models
3. Improved nullability support - fewer extra null checks, correct nullable types generated. Might require adding missing annotations to the swagger spec.
4. Improved header support - descriptions and `x-ms-client-name` honored.
5. Single value, `modelAsString` enums are generated as types, they were string constants before.

What you might need to do:
1. If you copied and overrode the serialization code you might want to update it to be in sync with new generator patterns.
2. **Please follow up on updating the centralized swagger file if there were workarounds applied to autorest.md**

### CODE REVIEW

File: sdk/digitaltwins/Azure.ResourceManager.DigitalTwins/api/Azure.ResourceManager.DigitalTwins.netstandard2.0.cs [hunk 1]
```diff
@@ -175,13 +175,30 @@ public partial class DigitalTwinsResource
         public string Location { get { throw null; } set { } }
         public string Name { get { throw null; } }
         public Azure.ResourceManager.DigitalTwins.Models.DigitalTwinsSkuInfo Sku { get { throw null; } set { } }
-        public System.Collections.Generic.IDictionary<string, string> Tags { get { throw null; } set { } }
+        public System.Collections.Generic.IDictionary<string, string> Tags { get { throw null; } }
         public string Type { get { throw null; } }
     }
+    [System.Runtime.InteropServices.StructLayoutAttribute(System.Runtime.InteropServices.LayoutKind.Sequential)]
```
Comment (drwill-ms) [inline sdk/digitaltwins/Azure.ResourceManager.DigitalTwins/api/Azure.ResourceManager.DigitalTwins.netstandard2.0.cs:181]: Do we want this added?

File: sdk/digitaltwins/Azure.DigitalTwins.Core/src/Generated/DigitalTwinModelsRestClient.cs [hunk 2]
```diff
@@ -88,14 +88,7 @@ public async Task<Response<PagedModelDataCollection>> ListAsync(IEnumerable<stri
                     {
                         PagedModelDataCollection value = default;
                         using var document = await JsonDocument.ParseAsync(message.Response.ContentStream, default, cancellationToken).ConfigureAwait(false);
-                        if (document.RootElement.ValueKind == JsonValueKind.Null)
```
Comment (drwill-ms) [inline sdk/digitaltwins/Azure.DigitalTwins.Core/src/Generated/DigitalTwinModelsRestClient.cs:91]: We have this same pattern copied into our customized code. Will this PR update those as well?

example: https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/digitaltwins/Azure.DigitalTwins.Core/src/Customized/Models/ModelData.Serialization.cs#L28

Comment (pakrym) [reply sdk/digitaltwins/Azure.DigitalTwins.Core/src/Generated/DigitalTwinModelsRestClient.cs:91]: No, the generator won't update the manually copied code. Having an extra null check doesn't hurt but as part of tightening the nullability rules we tried to keep them to a minimum in the generated code.

**Comment by drwill-ms** [comment]
Does this fall under #10295 
https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/digitaltwins/Azure.DigitalTwins.Core/src/Customized/Models/ModelData.cs#L54

File: sdk/digitaltwins/Azure.ResourceManager.DigitalTwins/api/Azure.ResourceManager.DigitalTwins.netstandard2.0.cs (hunk 1 above)
Comment (pakrym) [reply sdk/digitaltwins/Azure.ResourceManager.DigitalTwins/api/Azure.ResourceManager.DigitalTwins.netstandard2.0.cs:181]: The attribute is autogenerated by the compiler. We tend to keep them in the listings.

**Comment by drwill-ms** [comment]
For #3

> Improved nullability support - fewer extra null checks, correct nullable types generated. Might require adding missing annotations to the swagger spec.

Is this something our swagger owner should look at?

**Comment by pakrym** [comment]
> Is this something our swagger owner should look at?

I would add a temporary workaround to autorest.md but you should follow up with updating the central swagger.

**Comment by pakrym** [comment]
>Does this fall under # 2
>master/sdk/digitaltwins/Azure.DigitalTwins.Core/src/Customized/Models/ModelData.cs#L54

I think so.

File: sdk/digitaltwins/Azure.ResourceManager.DigitalTwins/src/autorest.md [hunk 3]
```diff
@@ -11,3 +11,15 @@ namespace: Azure.ResourceManager.DigitalTwins
 require: https://raw.githubusercontent.com/Azure/azure-rest-api-specs/8ba18e990a5c1a98d505018329d53272ce43f334/specification/digitaltwins/resource-manager/readme.md
 
 ```
+
+## Swagger workarounds
+
+### Add nullable annotations
+
+``` yaml
+directive:
```
Comment (pakrym) [inline sdk/digitaltwins/Azure.ResourceManager.DigitalTwins/src/autorest.md:20]: FYI

File: sdk/digitaltwins/Azure.DigitalTwins.Core/src/Generated/DigitalTwinModelsRestClient.cs (hunk 2 above)
Comment (drwill-ms) [reply sdk/digitaltwins/Azure.DigitalTwins.Core/src/Generated/DigitalTwinModelsRestClient.cs:91]: So we should update our customized code to match when we have the opportunity?

**Comment by drwill-ms** [comment]
> > Is this something our swagger owner should look at?
> 
> I would add a temporary workaround to autorest.md but you should follow up with updating the central swagger.

Note there are two SDKs you've updated in this PR. There is only one autorest.md file updated, I think.

File: sdk/digitaltwins/Azure.DigitalTwins.Core/src/Generated/DigitalTwinModelsRestClient.cs (hunk 2 above)
Comment (pakrym) [reply sdk/digitaltwins/Azure.DigitalTwins.Core/src/Generated/DigitalTwinModelsRestClient.cs:91]: Yes. I'll add this to the description that I use for these PRs.

**Review by drwill-ms — APPROVED** [review]

**Comment by pakrym** [comment]
>Note there are two SDKs you've updated in this PR. There is only one autorest.md file updated, I think.

autorest.md needs be updated only in the case where a swagger workaround is required.

**Pull request merged**